### PR TITLE
feat: add new package with vue extractor

### PR DESCRIPTION
**Pull request by j4hr3n**
# Description

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to Lingui?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

Fixes # (issue)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/lingui/js-lingui/blob/main/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/lingui/js-lingui/blob/main/CODE_OF_CONDUCT.md) docs
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/extractor-vue/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "repository": "lingui/js-lingui",
+  "bugs": "https://github.com/lingui/js-lingui/issues",
+  "license": "MIT",
+  "author": {
+    "name": "Christoffer Jahren",
+    "email": "christoffer@jahren.it"
+  },
+  "scripts": {
+    "build": "unbuild"
+  },
```
Comment (thekip) [inline packages/extractor-vue/package.json:26]: please copy scripts from other packages, including stub and cleaning

File: packages/extractor-vue/src/vue-extractor.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      .map((ext) => ext.slice(1))
+      .join("|") +
+    ")$",
+  "i"
+)
+
+export const vueExtractor: ExtractorType = {
+  match(filename: string) {
+    return babelRe.test(filename)
+  },
```
Comment (thekip) [inline packages/extractor-vue/src/vue-extractor.ts:19]: I think you can implement this extractor strictly for vue files and left rest for the original extractor: 

```ts
match(filename: string) {
  return filename.endsWith(".vue")
},
```

```ts
import { vueExtractor } from "@lingui/extractor-vue"
import babel from "@lingui/cli/api/extractors/babel"

const linguiConfig = {
  locales: ["en", "nb"],
  sourceLocale: "en",
  catalogs: [
    {
      path: "<rootDir>/src/{locale}",
      include: ["<rootDir>/src"],
    },
  ],
  extractors: [babel, vueExtractor],
}
export default linguiConfig
```

File: .gitignore [hunk 3]
```diff
@@ -2,6 +2,7 @@ node_modules/
 .yalc/
 yalc.lock
 dist
+build
```
Comment (thekip) [inline .gitignore:5]: intentionally?

**Review by thekip — CHANGES_REQUESTED** [review]

File: packages/extractor-vue/src/vue-extractor.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import * as vueCompiler from "@vue/compiler-sfc"
+// @ts-expect-error issue in typings, will be fixed shortly
+import babel from "@lingui/cli/api/extractors/babel"
```
[the remark this answers is not in the document]
Comment (j4hr3n) [reply packages/extractor-vue/src/vue-extractor.ts:3]: > that should not be an issue during development in monorepo.

@thekip I get this issue when building without it:

```
src/vue-extractor.ts(2,19): error TS2307: Cannot find module '@lingui/cli/api/extractors/babel' or its corresponding type declarations.
```

File: .gitignore (hunk 3 above)
Comment (j4hr3n) [reply .gitignore:5]: @thekip Yeah. When I run e.g `yarn release:build` locally to test if builds work it creates a build folder 🤷 I can remove it if you'd like.

Comment (thekip) [reply .gitignore:5]: Where exactly this folder is created?

File: packages/extractor-vue/src/vue-extractor.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    code: string,
+    onMessageExtracted,
+    ctx: ExtractorCtx
+  ) {
+    const {
+      descriptor: { script, scriptSetup },
+    } = vueCompiler.parse(code, { sourceMap: true, filename })
+
+    const finalCode = `${script?.content} ${scriptSetup?.content}`
+    const sourcemaps = `${script?.map} ${scriptSetup?.map}`
```
Comment (thekip) [inline packages/extractor-vue/src/vue-extractor.ts:21]: im not sure this is going to work like that

You actually can run `babel.extract` few times on each virtual file extracted from `.vue` if you don't want to "play" with sourcemaps.

File: packages/extractor-vue/src/vue-extractor.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  extract(
+    filename: string,
+    code: string,
+    onMessageExtracted,
+    ctx: ExtractorCtx
+  ) {
+    const {
+      descriptor: { script, scriptSetup },
+    } = vueCompiler.parse(code, { sourceMap: true, filename })
```
Comment (thekip) [inline packages/extractor-vue/src/vue-extractor.ts:18]: What was the issue to add a template for extraction as well?

File: packages/extractor-vue/src/vue-extractor.ts (hunk 5 above)
Comment (j4hr3n) [reply packages/extractor-vue/src/vue-extractor.ts:21]: > im not sure this is going to work like that
> 
> You actually can run `babel.extract` few times on each virtual file extracted from `.vue` if you don't want to "play" with sourcemaps.

@thekip Yeah will probably have to do something like that to make sourcemaps work

File: packages/extractor-vue/src/vue-extractor.ts (hunk 6 above)
Comment (j4hr3n) [reply packages/extractor-vue/src/vue-extractor.ts:18]: > What was the issue to add a template for extraction as well?

Could probably make that work as well. It's just a bit more intricate than the script portion of vue files.

Comment (j4hr3n) [reply packages/extractor-vue/src/vue-extractor.ts:18]: Also not sure if that is good practice/works in Vue. I'm not really a Vue developer myself, so my knowledge on that area is a bit lacking.

Comment (thekip) [reply packages/extractor-vue/src/vue-extractor.ts:18]: https://github.com/formatjs/formatjs/blob/main/packages/cli-lib/src/vue_extractor.ts#L72-L95

File: packages/extractor-vue/package.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  ],
+  "repository": "lingui/js-lingui",
+  "bugs": "https://github.com/lingui/js-lingui/issues",
+  "license": "MIT",
+  "author": {
+    "name": "Christoffer Jahren",
+    "email": "christoffer@jahren.it"
+  },
+  "scripts": {
+    "build": "unbuild",
```
Comment (thekip) [inline packages/extractor-vue/package.json:25]:
```suggestion
    "build": "rimraf ./dist && unbuild",
```

File: packages/extractor-vue/package.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@lingui/extractor-vue",
+  "version": "4.0.0-next.4",
```
Comment (thekip) [inline packages/extractor-vue/package.json:3]:
```suggestion
  "version": "4.0.0-next.5",
```

File: packages/extractor-vue/package.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  "scripts": {
+    "build": "unbuild",
+    "stub": "unbuild --stub"
+  },
+  "engines": {
+    "node": ">=16.0.0"
+  },
+  "dependencies": {
+    "@babel/core": "^7.21.0",
```
Comment (thekip) [inline packages/extractor-vue/package.json:32]: don'see `@babel/core` used in the code

File: packages/extractor-vue/src/__test__/extractor.test.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ * issues with snapshot tests failing on CI
+ * @param snapshot
+ */
+function removeOriginFromSnapshot(snapshot: ExtractedMessage[]) {
+  snapshot.forEach((entry) => {
+    if ("origin" in entry) {
+      delete entry.origin
+    }
+  })
+}
```
Comment (thekip) [inline packages/extractor-vue/src/__test__/extractor.test.ts:18]: origins, and working with sourcemaps is an important part of the extractor. You could not just ignore the, in tests. If you have problems with windows path or with diffrent root in CI, it's better to normalize them (cut path from root)

File: packages/extractor-vue/src/utils/regex-extractor.ts
Comment (thekip) [inline packages/extractor-vue/src/utils/regex-extractor.ts:1]: it's wrong way. Parsing any code with regex would never succeed.

File: packages/extractor-vue/pnpm-lock.yaml
Comment (thekip) [inline packages/extractor-vue/pnpm-lock.yaml:1]: pnpm lock?

File: packages/extractor-vue/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { vueScriptExtractor } from "./vue-script-extractor"
+import { vueScriptSetupExtractor } from "./vue-script-setup-extractor"
+import { vueTemplateExtractor } from "./vue-template-extractor"
+
+export const vueExtractors = [
+  vueScriptExtractor,
+  vueScriptSetupExtractor,
+  vueTemplateExtractor,
+]
```
Comment (thekip) [inline packages/extractor-vue/src/index.ts:9]: this way you will parse vue files with vue-sfc-compiler 3 times...

Comment (j4hr3n) [reply packages/extractor-vue/src/index.ts:9]: > this way you will parse vue files with vue-sfc-compiler 3 times...

Happy for you to take a stab on this. Not sure how to make it work properly with sourcemaps and all by merging handling of all Vue SFC parts into one extractor.

File: packages/extractor-vue/src/utils/regex-extractor.ts
Comment (j4hr3n) [reply packages/extractor-vue/src/utils/regex-extractor.ts:1]: > it's wrong way. Parsing any code with regex would never succeed.

Not sure how to solve it any other way tbh. Would need some help/guidance to make this work properly then.

**Comment by thekip** [comment]
Fully working extractor for vue, including templates: 
```ts
import { parse, compileTemplate, SFCBlock } from "@vue/compiler-sfc"
import babel from "@lingui/cli/api/extractors/babel"
import type { ExtractorCtx, ExtractorType } from "@lingui/conf"

export const vueScriptExtractor: ExtractorType = {
  match(filename: string) {
    return filename.endsWith(".vue")
  },
  async extract(
    filename: string,
    code: string,
    onMessageExtracted,
    ctx: ExtractorCtx
  ) {
    const { descriptor } = parse(code, {
      sourceMap: true,
      filename,
      ignoreEmpty: true,
    })

    const compiledTpl = compileTemplate({
      source: code,
      filename,
      // i don't know what it for, but it's required
      id: "id",
    })

    const isTsBlock = (block: SFCBlock) => block?.attrs?.lang === "ts"

    const targets = [
      [
        descriptor.script?.content,
        descriptor.script?.map,
        isTsBlock(descriptor.script),
      ],
      [
        descriptor.scriptSetup?.content,
        descriptor.scriptSetup?.map,
        isTsBlock(descriptor.script),
      ],
      [compiledTpl?.code, compiledTpl?.map, isTsBlock(descriptor.script)],
    ] as const

    await Promise.all(
      targets
        .filter(([source]) => Boolean(source))
        .map(([source, map, isTs]) =>
          babel.extract(
            filename + (isTs ? ".ts" : ""),
            source,
            onMessageExtracted,
            {
              sourceMaps: map,
              ...ctx,
            }
          )
        )
    )
  },
}
```

// example-template.vue
```vue
<script lang="ts">
import { i18n } from "@lingui/core"
import { defineComponent } from "vue"

export default defineComponent({
  data() {
    return {
      i18n,
    }
  },
})
</script>

<template>
  {{ i18n.t(/*i18n*/{ id: "custom.id", message: "My message" }) }}
  {{
    i18n.t(/*i18n*/{
      message: "My awesome message",
      id: "awesome.message",
      comment: "Yeah quite awesome huh?",
    })
  }}
  {{ i18n._(/*i18n*/"id used as message") }}
  {{ i18n.t(/*i18n*/{ id: "My message without ID and context" }) }}
</template>
```
As i mentioned earlier, there no reliable way to detect subjects for extracting other than manual annotating. When template is compiled all `i18n._` calls become `ctx.i18n._`.

File: packages/extractor-vue/src/vue-extractor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (thekip) [reply packages/extractor-vue/src/vue-extractor.ts]: Add 

```json
      "@lingui/cli/api/extractors/babel": ["./packages/cli/src/api/extractors/babel"],
```

to tsconfig.json in path mapping

**Comment by j4hr3n** [comment]
> Fully working extractor for vue, including templates:
> 
> ```ts
> import { parse, compileTemplate, SFCBlock } from "@vue/compiler-sfc"
> import babel from "@lingui/cli/api/extractors/babel"
> import type { ExtractorCtx, ExtractorType } from "@lingui/conf"
> 
> export const vueScriptExtractor: ExtractorType = {
>   match(filename: string) {
>     return filename.endsWith(".vue")
>   },
>   async extract(
>     filename: string,
>     code: string,
>     onMessageExtracted,
>     ctx: ExtractorCtx
>   ) {
>     const { descriptor } = parse(code, {
>       sourceMap: true,
>       filename,
>       ignoreEmpty: true,
>     })
> 
>     const compiledTpl = compileTemplate({
>       source: code,
>       filename,
>       // i don't know what it for, but it's required
>       id: "id",
>     })
> 
>     const isTsBlock = (block: SFCBlock) => block?.attrs?.lang === "ts"
> 
>     const targets = [
>       [
>         descriptor.script?.content,
>         descriptor.script?.map,
>         isTsBlock(descriptor.script),
>       ],
>       [
>         descriptor.scriptSetup?.content,
>         descriptor.scriptSetup?.map,
>         isTsBlock(descriptor.script),
>       ],
>       [compiledTpl?.code, compiledTpl?.map, isTsBlock(descriptor.script)],
>     ] as const
> 
>     await Promise.all(
>       targets
>         .filter(([source]) => Boolean(source))
>         .map(([source, map, isTs]) =>
>           babel.extract(
>             filename + (isTs ? ".ts" : ""),
>             source,
>             onMessageExtracted,
>             {
>               sourceMaps: map,
>               ...ctx,
>             }
>           )
>         )
>     )
>   },
> }
> ```
> 
> // example-template.vue
> 
> ```
> <script lang="ts">
> import { i18n } from "@lingui/core"
> import { defineComponent } from "vue"
> 
> export default defineComponent({
>   data() {
>     return {
>       i18n,
>     }
>   },
> })
> </script>
> 
> <template>
>   {{ i18n.t(/*i18n*/{ id: "custom.id", message: "My message" }) }}
>   {{
>     i18n.t(/*i18n*/{
>       message: "My awesome message",
>       id: "awesome.message",
>       comment: "Yeah quite awesome huh?",
>     })
>   }}
>   {{ i18n._(/*i18n*/"id used as message") }}
>   {{ i18n.t(/*i18n*/{ id: "My message without ID and context" }) }}
> </template>
> ```
> 
> As i mentioned earlier, there no reliable way to detect subjects for extracting other than manual annotating. When template is compiled all `i18n._` calls become `ctx.i18n._`.

@thekip Awesome! 👏  Let me test it out 😄

File: packages/extractor-vue/src/vue-extractor.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
+import { parse, compileTemplate, SFCBlock } from "@vue/compiler-sfc"
+// @ts-expect-error temporary typing issue
+import babel from "@lingui/cli/api/extractors/babel"
```
Comment (thekip) [inline packages/extractor-vue/src/vue-extractor.ts:3]: add `"@lingui/cli/api/extractors/babel": ["./packages/cli/src/api/extractors/babel"],` to root paths in `tsconfig.json`

**Review by thekip — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (thekip) [inline .gitignore:5]: it's not resolved

File: packages/extractor-vue/src/__test__/test.vue [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+<script lang="ts">
+export default defineComponent({
+  data() {
+    return {
+      i18n,
+      scriptString: i18n._("Script message"),
+    }
+  },
+})
+</script>
```
Comment (thekip) [inline packages/extractor-vue/src/__test__/test.vue:17]: add specific to TS only syntax, to validate that extractor will not fail with "unexpected token",  something simple, for example: 

```ts
const foo: Type = 5;
```

File: packages/extractor-vue/src/test.vue
Comment (thekip) [inline packages/extractor-vue/src/test.vue:1]: still not the right place for this file. You either create a `test` folder _next to_ `src` and put test file + fixtures + snapshots there. Or create folder `src/fixtures` and put `*.vue` file there. 

Check `format-po` for reference

**Comment by j4hr3n** [comment]
@thekip updated with your patch now!

**Comment by thekip** [comment]
please rebase to the latest `next` branch, and update readme in a package. Thanks to https://github.com/lingui/js-lingui/pull/1586 it should now extract from template from all type of i18n calls without explicit annotations. (it also would be nice to update a fixture as well)

**Comment by j4hr3n** [comment]
@thekip updated now

File: packages/extractor-vue/README.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
+# Vue extractor
+
+This package contains a set of custom extractors that handles Vue files. It supports extracting messages from script and setup scripts as well as limited support for extracting messages from templates.
```
Comment (thekip) [inline packages/extractor-vue/README.md:3]: it's not limited now

**Review by thekip — APPROVED** [review]

**Comment by thekip** [comment]
@andrii-bodnar

File: packages/extractor-vue/README.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
+# Vue extractor
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:1]: please follow the same Readme structure as in other lingui packages

File: packages/extractor-vue/README.md [hunk 16]
```diff
@@ -30,18 +36,16 @@ const linguiConfig = {
...
-While the following examples wont:
+This package is licensed under [MIT][license] license.
 
-- `i18n._("Message")`
-- `i18n.t({ id: "Message" })`
-- `i18n.t({ message: "Message", id: "my.message", comment: "Comment" })`
+[license]: https://github.com/lingui/js-lingui/blob/main/LICENSE
+[linguijs]: https://github.com/lingui/js-lingui
+[documentation]: https://lingui.dev
+[tutorial]: https://lingui.dev/tutorials/cli
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:46]: unused

File: packages/extractor-vue/README.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+# @lingui/vue-extractor
+
+This package contains a custom extractor that handles Vue files. It supports extracting messages from script and setup scripts as well Vue templates.
+
+`@lingui/vue-extractor` is part of [LinguiJS][linguijs]. See the [documentation][documentation] for all information, tutorials and examples.
+
+## Installation
+
+```sh
+npm install @lingui/extractor-vue
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:14]: `--save-dev`?

**Comment by andrii-bodnar** [comment]
@j4hr3n It would be great also to have the website page in the "API Reference" or "Guide" section about this extractor so users will be able to easily find it

File: packages/extractor-vue/package.json [hunk 18]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@lingui/extractor-vue",
+  "version": "4.0.0-next.5",
+  "description": "Custom vue extractor to be used with CLI tool",
```
Comment (andrii-bodnar) [inline packages/extractor-vue/package.json:4]:
```suggestion
  "description": "Custom Vue.js extractor to be used with the CLI tool",
```

File: packages/extractor-vue/README.md [hunk 19]
```diff
@@ -0,0 +1,51 @@
+[![License][badge-license]][license]
+[![Version][badge-version]][package]
+[![Downloads][badge-downloads]][package]
+
+# @lingui/vue-extractor
+
+This package contains a custom extractor that handles Vue files. It supports extracting messages from script and setup scripts as well Vue templates.
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:7]:
```suggestion
This package contains a custom extractor that handles Vue.js files. It supports extracting messages from script and setup scripts as well as Vue templates.
```

File: packages/extractor-vue/README.md [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+## Installation
+
+```sh
+npm install @lingui/extractor-vue
+```
+
+## Usage
+
+This custom extractor requires that you use javascript for your lingui configuration.
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:19]:
```suggestion
This custom extractor requires that you use JavaScript for your Lingui configuration.
```

Comment (thekip) [reply packages/extractor-vue/README.md:19]: I'm thinking would it be better to write here something similar as we do in docs:

> To use custom extractor add to your `lingui.config.{js,ts}`

**Comment by andrii-bodnar** [comment]
@j4hr3n could you please rebase the branch and address my latest comment about documenting this extractor?

**Comment by j4hr3n** [comment]
> @j4hr3n could you please rebase the branch and address my latest comment about documenting this extractor?

@andrii-bodnar Yes I'll update this shortly

**Comment by j4hr3n** [comment]
@andrii-bodnar website page has been added now

File: website/docs/ref/extractor-vue.md [hunk 21]
```diff
@@ -0,0 +1,32 @@
+# Vue.js extractor
+
+The `@lingui/extractor-vue` package provides a [custom extractor](../guides/custom-extractor.md) that handles Vue.js files.
```
Comment (andrii-bodnar) [inline website/docs/ref/extractor-vue.md:3]:
```suggestion
The `@lingui/extractor-vue` package provides a [custom extractor](/docs/guides/custom-extractor.md) that handles Vue.js files.
```

Comment (andrii-bodnar) [reply website/docs/ref/extractor-vue.md:3]: also, I would suggest moving this article to the tutorials section + adding the "## Further reading" section with a link to the "Custom Extractor" article

File: website/docs/tutorials/extractor-vue.md [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+  extractors: [babel, vueExtractor],
+}
+
+export default linguiConfig
+```
+
+## Further reading
+
+- [Message extraction](/docs/guides/message-extraction.md)
+- [Custom extractors](/docs/guides/custom-extractor.md)
```
Comment (andrii-bodnar) [inline website/docs/tutorials/extractor-vue.md:37]:
```suggestion
- [Message Extraction](/docs/guides/message-extraction.md)
- [Custom Extractors](/docs/guides/custom-extractor.md)
```

File: website/docs/tutorials/extractor-vue.md [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+```bash npm2yarn
+npm install --save-dev @lingui/extractor-vue
+```
+
+## Usage
+
+It is required that you use JavaScript for your Lingui configuration.
+
+```js
```
Comment (andrii-bodnar) [inline website/docs/tutorials/extractor-vue.md:15]:
```suggestion
```js title="lingui.config.{ts,js}"
```

File: packages/extractor-vue/README.md [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+export default linguiConfig
+```
+
+## License
+
+This package is licensed under [MIT][license] license.
+
+[license]: https://github.com/lingui/js-lingui/blob/main/LICENSE
+[linguijs]: https://github.com/lingui/js-lingui
+[documentation]: https://lingui.dev
```
Comment (andrii-bodnar) [inline packages/extractor-vue/README.md:46]: Please add the direct tutorial link here

File: packages/extractor-vue/package.json [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+  "scripts": {
+    "build": "rimraf ./dist && unbuild",
+    "stub": "unbuild --stub"
+  },
+  "engines": {
+    "node": ">=16.0.0"
+  },
+  "dependencies": {
+    "@lingui/cli": "^4.0.0-next.5",
+    "@lingui/conf": "^4.0.0-next.5",
```
Comment (andrii-bodnar) [inline packages/extractor-vue/package.json:34]:
```suggestion
    "@lingui/cli": "^4.0.0-next.7",
    "@lingui/conf": "^4.0.0-next.7",
```

File: packages/extractor-vue/package.json [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+    "@vue/compiler-sfc": "^3.2.47"
+  },
+  "devDependencies": {
+    "@lingui/babel-plugin-extract-messages": "^4.0.0-next.5",
```
Comment (andrii-bodnar) [inline packages/extractor-vue/package.json:38]:
```suggestion
    "@lingui/babel-plugin-extract-messages": "^4.0.0-next.7",
```

File: packages/extractor-vue/package.json [hunk 27]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@lingui/extractor-vue",
+  "version": "4.0.0-next.5",
```
Comment (andrii-bodnar) [inline packages/extractor-vue/package.json:3]:
```suggestion
  "version": "4.0.0-next.7",
```

**Comment by andrii-bodnar** [comment]
@j4hr3n please address my latest comments and then this PR will be merged

**Comment by j4hr3n** [comment]
> @j4hr3n please address my latest comments and then this PR will be merged

@andrii-bodnar done

**Review by andrii-bodnar — APPROVED** [review]

**Comment by Touffy** [comment]
I managed to make this work on a poorly-maintained Vue 2.6 project by fixing two things :

adding `. default` when importing the base extractor
```js
const babel = require('@lingui/cli/api/extractors/babel').default;
```

and handling the case of an SFC with no `<template>`
```js
    const compiledTemplate = descriptor.template && compilerSfc.compileTemplate({
      source: descriptor.template.content,
      filename,
      inMap: descriptor.template.map,
      id: filename
    });
```

**Comment by Touffy** [comment]
Sure.

**Comment by Touffy** [comment]
https://github.com/lingui/js-lingui/pull/1623

I don't know how to change how the `import` is transpiled to `require`.

**Comment by thekip** [comment]
you don't have to, the `default` export is painful, depending on the consumer environment it will give different result.

If you consuming from pure ESM or using transpiling with interop and etc. 

In a newer code i'm avoiding `default` exports as much as possible, but this extractor is being there for a long time.

So if with additional `.default` it works for you, just leave it as it is in your code. No other changes needed.

**Pull request merged**